Source/DafnyLanguageServer.Test/Lookup/ReferencesTest.cs

```diff
@@ -0,0 +1,278 @@
+    /// Assert that when finding-references at each cursor position,
+    /// the client returns all ranges marked with regular spans.
+    /// </summary>
+    /// <param name="source"></param>
+    private async Task AssertReferences(string source) {
+      MarkupTestFile.GetPositionsAndRanges(
+        source, out var cleanSource, out var positions, out var expectedRangesArray);
+      var expectedRanges = new HashSet<Range>(expectedRangesArray);
+
+      var documentItem = CreateTestDocument(cleanSource);
+      await client.OpenDocumentAndWaitAsync(documentItem, CancellationToken);
+
+      foreach (var position in positions) {
+        var result = await RequestReferences(documentItem, position).AsTask();
+        var resultRanges = result.Select(location => location.Range).ToHashSet();
+        Assert.Equal(expectedRanges, resultRanges);
+      }
+    }
+
+    [Fact]
+    public async Task Const() {
+      var source = @"
+const ><c := 1
+method M1() {
+  print [>><c<];
+}
+method M2() {
+  print [>c><<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Var() {
+      var source = @"
+method M() {
+  var ><v := 1;
+  print [>><v<];
+  print [>><v<];
+  var v2 := [>><v<] + [>><v<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Method() {
+      var source = @"
+method ><M1() {
+}
+method M2() {
+  [>><M1<]();
+}
+method M3() {
+  [>><M1<]();
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Parameter() {
+      var source = @"
+method M(><x: int) {
+  print [>><x<];
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task Module() {
+      var source = @"
+module ><M1 {
+}
+module M2 {
+  import [>><M1<]
+}
+module M3 {
+  import [>><M1<]
+}
+module MR refines [>><M1<] {}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // It seems that datatype declaration/usage information is not tracked
+    [Fact(Skip = "Not implemented")]
+    public async Task DatatypeDeclaration() {
+      var source = @"
+datatype ><D = D
+method M(d: [>><D<]) {
+  print (match d
+    case D => 1
+  );
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task DatatypeConstructor() {
+      var source = @"
+datatype Letter = ><A | B | C
+method M(l: Letter) {
+  print match l
+    case [>><A<] => 1
+    case B => 2
+    case C => 3
+  ;
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    [Fact]
+    public async Task DatatypeDestructor() {
+      var source = @"
+datatype Option = None | Some(><value: int)
+method M(o: Option) {
+  print match o
+    case None => 0
+    case Some(value) => value + o.[>><value<]
+  ;
+}
+".TrimStart();
+
+      await AssertReferences(source);
+    }
+
+    // It seems that type declaration/usage information is not tracked
```

Comment:
Probably resolved by adding implementing `IHasUsages` on more AST nodes, but we can fix that in another PR

Reply:
I think that's a good idea, and I agree that it would be better to address in another PR.

On a related note, it would be nice if the declaration graph were reflexive on declarations, i.e. `GetDeclarations(myDecl.Position)` would include `myDecl` itself. This seems moreintuitive to me, and it would also simplify the implementation of `DafnyDefinitionHandler.Handle`.

Reply:
I don't follow. Instead of `GetDeclarations ` and `DafnyDefinitionHandler.Handle`, do you mean `GetUsages` and `DafnyReferencesHandler.Handle` ?

Note that LSP allows the client to request usages of a declaration that doesn't return the declaration itself:

<img width="653" alt="image" src="https://github.com/dafny-lang/dafny/assets/3121201/b3a8f958-089e-48cc-b6e5-9fc58fc6aeb5">

Reply:
Yes, I did mean `GetUsages` and `DafnyReferencesHandle.Handle`.

I'll implement support for `includeDeclaration` being false.